drivers: sensor: adiltc2990: add adi ltc2990 support

jilaypandya opened the pull request:
This PR adds basic support of analog devices LTC2990 Quad I2C Voltage, Current and Temperature Monitor.
This sensor was tested with MIMXRT1062xxx6A for monitoring 4 single ended voltages with following output.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/126553325/e494ed48-1c41-48c7-9332-aef27c9a836a)

MaureenHelm commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -2,6 +2,8 @@
 add_subdirectory_ifdef(CONFIG_A01NYUB a01nyub)
 add_subdirectory_ifdef(CONFIG_ADC_CMP_NPCX nuvoton_adc_cmp_npcx)
+add_subdirectory_ifdef(CONFIG_ADLTC2990 adltc2990)
+
```
Please remove the empty line

jilaypandya replied:
done

MaureenHelm commented on dts/bindings/sensor/adi,adltc2990.yaml:
```diff
@@ -0,0 +1,79 @@
+include: [sensor-device.yaml, i2c-device.yaml]
+
+properties:
+  reg:
+    required: true
+    description: |
+      i2c device address
```
This is redundant (already defined in included `i2c-device.yaml`), please remove.

jilaypandya replied:
done

MaureenHelm commented on dts/bindings/sensor/adi,adltc2990.yaml:
```diff
@@ -0,0 +1,79 @@
+      - 1
+      - 2
+      - 3
+      - 4
```
Please look at following the linux kernel bindings https://www.kernel.org/doc/Documentation/devicetree/bindings/hwmon/ltc2990.txt

jilaypandya replied:
Hi @MaureenHelm thanks for the detailed review, That was also my initial idea to have a combination of mode_4_3 and mode_2_0 control register values. However, I thought of simplifying it for the user to select just the physical value to be measured. In the approach from linux kernel bindings, a deeper understanding of the combination of the control register values is required. Would it be okay to leave the mode selection like this?

MaureenHelm commented on drivers/sensor/adltc2990/Kconfig:
```diff
@@ -0,0 +1,49 @@
+
+menuconfig ADLTC2990
+	bool "ADLTC2990 Quad I2C Voltage, Current and Temperature Monitor"
+	default n
```
```suggestion
	default y
```

jilaypandya replied:
done

MaureenHelm commented on drivers/sensor/adltc2990/Kconfig:
```diff
@@ -0,0 +1,49 @@
+
+if ADLTC2990
+
+choice
+	prompt "Temperature Format"
+
+config ADLTC2990_TEMPERATURE_IN_KELVIN
+	bool "Kelvin"
+
+config ADLTC2990_TEMPERATURE_IN_CELCIUS
+	bool "Celcius"
+
+endchoice
+
+config ADLTC2990_TEMPERATURE_FORMAT_VALUE
+	int
+	default 0 if ADLTC2990_TEMPERATURE_IN_CELCIUS
+	default 1 if ADLTC2990_TEMPERATURE_IN_KELVIN
+
+choice
+	prompt "Acquisition Type"
+
+config ADLTC2990_ACQUISITION_SINGLE
+	bool "Single"
+
+config ADLTC2990_ACQUISITION_REPEATED
+	bool "Repeated"
+
+endchoice
+
+config ADLTC2990_ACQUISITION_TYPE_VALUE
+	int
+	default 0 if ADLTC2990_ACQUISITION_REPEATED
+	default 1 if ADLTC2990_ACQUISITION_SINGLE
```
These correspond directly to hardware register fields; move to devicetree properties.

jilaypandya replied:
done

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,471 @@
+#define TEMP_VALUES                 1
+#define CURRENT_VALUES              1
+#define MICROOHM_CONVERSION_FACTOR  1000000
+static int adltc2990_bus_reg_read_i2c(const struct device *dev, uint8_t reg_address, uint8_t *value)
+{
+	const struct adltc2990_config *cfg = dev->config;
+
+	return i2c_reg_read_byte_dt(&cfg->bus, reg_address, value);
+}
+
+static int adltc2990_bus_reg_write_i2c(const struct device *dev, uint8_t reg_address, uint8_t value)
+{
+	const struct adltc2990_config *cfg = dev->config;
+
+	return i2c_reg_write_byte_dt(&cfg->bus, reg_address, value);
+}
```
These wrapper functions aren't necessary

jilaypandya replied:
done

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,471 @@
+		LOG_ERR("This configuration is not allowed");
+	} else if (mode_v1_v2 == nothing && mode_v3_v4 == nothing) {
+		conf_value = 0;
+	} else if (mode_v1_v2 == nothing &&
+		   (mode_v3_v4 == current || mode_v3_v4 == voltage_differential)) {
+		conf_value = 20;
+	} else if (mode_v1_v2 == nothing && mode_v3_v4 == voltage_single_ended) {
+		conf_value = 18;
+	} else if (mode_v1_v2 == nothing && mode_v3_v4 == temperature) {
+		conf_value = 16;
+	} else if ((mode_v1_v2 == current || mode_v1_v2 == voltage_differential) &&
+		   mode_v3_v4 == nothing) {
+		conf_value = 9;
+	} else if ((mode_v1_v2 == current || mode_v1_v2 == voltage_differential) &&
+		   (mode_v3_v4 == current || mode_v3_v4 == voltage_differential)) {
+		conf_value = 30;
+	} else if ((mode_v1_v2 == current || mode_v1_v2 == voltage_differential) &&
+		   mode_v3_v4 == voltage_single_ended) {
+		conf_value = 26;
+	} else if ((mode_v1_v2 == current || mode_v1_v2 == voltage_differential) &&
+		   mode_v3_v4 == temperature) {
+		conf_value = 25;
+	} else if (mode_v1_v2 == voltage_single_ended && mode_v3_v4 == nothing) {
+		conf_value = 8;
+	} else if (mode_v1_v2 == voltage_single_ended &&
+		   (mode_v3_v4 == current || mode_v3_v4 == voltage_differential)) {
+	} else if (mode_v1_v2 == voltage_single_ended && mode_v3_v4 == voltage_single_ended) {
+		conf_value = 31;
+	} else if (mode_v1_v2 == voltage_single_ended && mode_v3_v4 == temperature) {
+		conf_value = 24;
+	} else if (mode_v1_v2 == temperature && mode_v3_v4 == nothing) {
+		conf_value = 12;
+	} else if (mode_v1_v2 == temperature &&
+		   (mode_v3_v4 == current || mode_v3_v4 == voltage_differential)) {
+		conf_value = 28;
+	} else if (mode_v1_v2 == temperature && mode_v3_v4 == voltage_single_ended) {
+		conf_value = 11;
+	} else if (mode_v1_v2 == temperature && mode_v3_v4 == temperature) {
+		conf_value = 29;
+	}
```
There are way too many `else if` clauses; please try to simplify

jilaypandya replied:
done

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.h:
```diff
@@ -0,0 +1,60 @@
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/sensor.h>
+
+typedef enum adltc2990_monitor_pins {
```
Not sure why checkpatch didn't catch this, but you shouldn't add new typedefs

MaureenHelm reviewed (changes requested):
Thank you for your contribution!

Please combine commits `drivers: sensor: adds adltc2990 to CMakeLists.txt and Kconfig` and driver: sensor: adds basic support for analog devices ltc2990. You also need to add this sensor to `tests/drivers/build_all/sensor/i2c.dtsi` to build the driver in CI.

MaureenHelm commented on dts/bindings/sensor/adi,adltc2990.yaml:
```diff
@@ -0,0 +1,74 @@
+include: [sensor-device.yaml, i2c-device.yaml]
+
+properties:
+  mode_v1_v2:
```
Convention is to use hyphens instead of underscores for devicetree properties

MaureenHelm commented on dts/bindings/sensor/adi,adltc2990.yaml:
```diff
@@ -0,0 +1,74 @@
+  pins_v1_v2_current_resistor:
+    type: int
+    description: Define the resistor to be used for measuring current in microohms
+    default: 1000
```
Defaults need justification in the description. Is it a register reset value?

jilaypandya replied:
In order to measure the current from voltage differenece of V1 and V2, some resistor would be there. And this is that resistor. I have just added some default value of 1 milliohms, but what would be the best way to represent this resistor?

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,484 @@
+static enum adltc2990_monitoring_type adltc2990_get_v1_v2_measurement_modes(uint8_t mode_4_3,
+									    uint8_t mode_2_0)
+{
+	if (mode_2_0 > 7 || mode_4_3 > 3) {
+		LOG_ERR("Invalid Measurement Mode");
+		return -EINVAL;
+	}
+	if (mode_4_3 == 0 || mode_4_3 == 2) {
+		return nothing;
+	}
+
+	enum adltc2990_monitoring_type type = nothing;
+
+	switch (mode_2_0) {
+	case 0:
+	case 7:
+		type = voltage_single_ended;
+		break;
+
+	case 1:
+	case 2:
+	case 6:
+		type = voltage_differential;
+		break;
+
+	case 3:
+	case 4:
+	case 5:
```
Please use macros with descriptive names for these magic numbers

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,484 @@
+	uint8_t status_reg = 0;
+
+	i2c_reg_read_byte_dt(&cfg->bus, ADLTC2990_REG_STATUS, &status_reg);
+	return status_reg & 0b00000001;
```
```suggestion
	return status_reg & BIT(0);
```

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,484 @@
+		break;
+
+	default:
+		LOG_ERR("does not measure channel: %d", chan);
```
Return `-ENOTSUP`

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,484 @@
+		break;
+
+	default:
+		return -EINVAL;
```
`-ENOTSUP`

MaureenHelm commented on dts/bindings/sensor/adi,adltc2990.yaml:
```diff
@@ -9,22 +9,52 @@ compatible: "adi,adltc2990"
 
 properties:
-  mode_v1_v2:
+  temperature_format:
```
Please squash the binding changes into one commit so you don't introduce a new binding in one commit and immediately change them in the next commit.

jilaypandya replied:
my bad sorry

jilaypandya committed:
dts: bindings: adiltc2990: add dts bindings for analog devices ltc2990

This commit adds adi,adltc2990.yaml to dts/bindings/sensor.

Signed-off-by: Jilay Pandya <jilay.pandya@zeiss.com>

jilaypandya committed:
driver: sensor: adds basic support for analog devices ltc2990

This commit adds Kconfig variables to configure ADLTC2990. This commit adds
basic driver code for analog devices ltc2990 sensor.

Signed-off-by: Jilay Pandya <jilay.pandya@zeiss.com>

jilaypandya commented:
@MaureenHelm @yperess,
Hello Together :-),

I have incorporated the requested changes, hope the PR is fine now. Would be great if could get a further review or approval:-)

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,485 @@
+
+	const uint16_t val = msb + lsb;
+
+	const bool is_neg = (msb_value & negative_bit_index) ? true : false;
+	int32_t ret = is_neg ? ((~val & bitmask) + 1) * -1 : (val & bitmask);
```
Is this trying to sign extend? It would be simpler to left shift then right shift, and you can do that unconditionally without checking if the quantity is negative.

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,485 @@
+	const uint8_t ctrl_reg_setting = cfg->temp_format << 7 | cfg->acq_format << 6 | 0 << 5 |
+					 cfg->measurement_mode[1] << 3 | cfg->measurement_mode[0];
+
+	LOG_INF("Setting Control Register to: 0x%x", ctrl_reg_setting);
```
This is probably too noisy, use `LOG_DBG` instead.

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,485 @@
+	int err = i2c_reg_write_byte_dt(&cfg->bus, ADLTC2990_REG_CONTROL, ctrl_reg_setting);
+
+	if (err < 0) {
+		LOG_DBG("configuring for single bus failed: %d", err);
```
This can be `LOG_ERR`

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,485 @@
+	case SENSOR_CHAN_DIE_TEMP: {
+		val->val1 = data->internal_temperature / 10000;
+		val->val2 = data->internal_temperature % 10000;
+		LOG_INF("Internal Temperature Value is:%d.%d", val->val1, val->val2);
```
This may get noisy too, use `LOG_DBG` for all the successful cases in this function

MaureenHelm commented on drivers/sensor/adltc2990/adltc2990.h:
```diff
@@ -0,0 +1,63 @@
+	uint8_t temp_format;
+	uint8_t acq_format;
+	uint8_t measurement_mode[2];
+	struct pins_configuration pins_v1_v2_configuration;
+	struct pins_configuration pins_v3_v4_configuration;
```
```suggestion
	struct pins_configuration pins_v1_v2;
	struct pins_configuration pins_v3_v4;
```
The `_configuration` suffix is redundant and dropping it will simplify the driver a bit.

MaureenHelm reviewed (changes requested):
This is looking better, thank you!

jilaypandya commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,470 @@
+	int16_t value = (msb_value<<8)+lsb_value;
+
+	int32_t voltage_value = (value << (31-negative_bit_index))>>(31-negative_bit_index);
+
```
In response to : https://github.com/zephyrproject-rtos/zephyr/pull/60842#discussion_r1323581251
Hi @MaureenHelm, i guess this is how you meant it, or?

MaureenHelm replied:
Yes

ybkahveci commented on drivers/sensor/adltc2990/adltc2990.c:
```diff
@@ -0,0 +1,470 @@
+			LOG_DBG("Getting V1,V2");
+			num_values_v1_v2 = ADLTC2990_VOLTAGE_SINGLE_ENDED_VALUES;
+		} else if (mode_v1_v2 == VOLTAGE_DIFFERENTIAL) {
+			LOG_DBG("Getting V3-V4");
```
Should be 
`LOG_DBG("Getting V1-V2");`
